### PR TITLE
Further NXP P1021 fixes and improvements

### PR DESCRIPTION
**Pull request by dgarske**
* Added ELF32 and ELF64 loader support (config `ELF=1` or build option `WOLFBOOT_ELF`).
* Add ELF support to update_ram.c loader.
* Add support for loading entire flash image to RAM when `EXT_FLASH=1` and `NO_XIP=1` (or `WOLFBOOT_USE_RAMBOOT`). 
* Added QUICC Engine support to load microcode and enable.
* Add multiple core support for NXP P1021.
* Fixes to resolve first stage boot ROM relocation.
  - Implemented temporary workaround to resolve stack traps.
* Added GOT relocation support.
* Fix for the PPC `isr_empty` handler address.
* Fix to allow stack to use DDR by having assembly setup DDR TLB
* Cleanup wolfBoot output. 
  - Only remove extra .bin/.elf created unless `make distclean` is used.
  - Don't output the key grep test (only check result). 
  - Adjust build order (first stage, wolfboot, test app, key, sign test app and factory).
  - Fix to make sure linker script is rebuilt before objects.
* Use DDR as stack after relocating (space before where it is relocated) and moved relocated 4k first stage loader to 0x8000 (for additional stack).
* Fixes and improvements with NXP T2080 regression testing.
* Added required LAW's to startup assembly.
* Moved LAW definitions to header.
* Move isr_empty so other boot functions go in the boot section.

Sample NXP P1021 Output:

```
Relocating BOOT ROM to DDR
Loading wolfBoot to DDR
Jumping to full wolfBoot
wolfBoot HAL Init
Flash Init: Ret 0, ID 0x76207620
QE: Length 63732, Count 1
QE: uploading 'Microcode for P1021 r1.0' version 0.0.1
QE: Traps 0
MP: Starting core 2 (spin table 0xFFFFF240)
Versions: Boot 1, Update 0
Trying Boot partition at 0x200000
Loading header 512 bytes to 0x1DFFFE00
Loading image 3170724 bytes to 0x1E000000
Image size 3170724
Firmware Valid
Loading elf at 0x1E000000
Found valid elf32 (big endian)
Booting at 0x6000
```

### CODE REVIEW

File: hal/nxp_p1021.c [hunk 1]
```diff
@@ -20,26 +20,26 @@
...
     #define ENABLE_PCIE
     #define ENABLE_CPLD /* Board Configuration and Status Registers (BCSR) */
     #define ENABLE_CONF_IO
+    #define ENABLE_QE   /* QUICC Engine */
+    #define ENABLE_ESPI /* SPI for TPM */
+    #define ENABLE_MP   /* multi-core support */
+    #define ENABLE_IRQ
 
-    #define ENABLE_ESPI /* TPM */
+    /* TODO - Ethernet*/
```
Comment (danielinux) [inline hal/nxp_p1021.c:40]: Remove "TODO" point or expand with a more verbose comment

File: hal/nxp_p1021.c [hunk 2]
```diff
@@ -906,6 +1010,20 @@ static int hal_cpld_init(void)
...
         MAS2_I | MAS2_G, 0, BOOKE_PAGESZ_256K, 1);
+
+    /* setup eLBC for CPLD (CS1), 8-bit */
+    set32(ELBC_BR(1), (ELBC_BR_ADDR(BCSR_BASE) | ELBC_BR_MSEL(0) |
+        ELBC_BR_PS(1) | ELBC_BR_V));
+    set32(ELBC_OR(1), (ELBC_OR_AMASK(ELBC_AMASK_32KB) |
+        ELBC_ORG_CSCT | ELBC_ORG_XACS | ELBC_ORG_SCY | ELBC_ORG_TRLX |
+        ELBC_ORG_EHTR | ELBC_ORG_EAD));
+
+#if 0 /* ethernet (uec) setup not required */
```
Comment (danielinux) [inline hal/nxp_p1021.c:1021]: Remove dead code or add another define to enable ethernet initialization from the config when appropriate

File: hal/nxp_p1021.c [hunk 3]
```diff
@@ -1058,26 +1176,421 @@ static void hal_io_init(void)
...
+        calc_size += sizeof(uint32_t) * firmware->microcode[i].count;
+    }
+
+    /* Validate the length */
+    if (length != calc_size + sizeof(uint32_t)) {
+        wolfBoot_printf("QE length %d invalid!\n", length);
+        return -1;
+    }
+
+#if 0 /* Disabled for wolfBoot */
```
Comment (danielinux) [inline hal/nxp_p1021.c:1315]: Remove dead code or add a config define to enforce CRC checking when appropriate

File: hal/nxp_p1021.c [hunk 4]
```diff
@@ -1058,26 +1176,421 @@ static void hal_io_init(void)
...
+        QE_IRAM_IADD_AIE | QE_IRAM_IADD_BADDR);
+
+    /* Copy 32-bits at a time to iRAM */
+    for (i = 0; i < ucode->count; i++) {
+        set32(QE_IRAM_IDATA, code[i]);
+    }
+}
+
+/* Upload a microcode to the I-RAM at a specific address */
+int qe_upload_firmware(const struct qe_firmware *firmware)
```
Comment (danielinux) [inline hal/nxp_p1021.c:1270]: should this be static? It does not seem to be used outside of this module

File: hal/nxp_p1021.c [hunk 5]
```diff
@@ -1058,26 +1176,421 @@ static void hal_io_init(void)
...
+    }
+#endif
+
+    /* start core and wait for it to be enabled */
+    hal_mp_up(bootpg);
+}
+#endif /* ENABLE_MP */
+
+#ifdef ENABLE_IRQ
+void hal_irq_init(void)
```
Comment (danielinux) [inline hal/nxp_p1021.c:1531]: static?

**Review by danielinux — CHANGES_REQUESTED** [review]
minor comments, mostly on dead code and symbols visibility

**Review by danielinux — COMMENTED** [review]
Good to go. Please consider making hal_irq_init, hal_l2_init static

**Comment by dgarske** [comment]
> Almost good to go. Please consider making hal_irq_init, hal_l2_init and other local functions static. Please expand the comment on the /* TODO */ regarding Ethernet mapping (i.e. why it's disabled). Please remove dead code (Ethernet initialization and CRC).

Hey Daniele,
I wanted to keep those non static so it would be possible to cal them individual

**Comment by dgarske** [comment]
> 

Done. I kept the QE CRC code in case someone has crc32 function and wants to enable it later.

File: hal/nxp_p1021.c (hunk 3 above)
Comment (danielinux) [reply hal/nxp_p1021.c]: OK if you want to keep this, but please change `#if 0` into something meaninful like `#ifdef QE_CRC_VALIDATE` even if not present in the current config

**Review by danielinux — CHANGES_REQUESTED** [review]
1- OK keeping the CRC code but please change if 0 into something more meaningful
2- Eth initialization: there is a "commented out" memcpy under `#if 0`, please use a meaningful define or eliminate dead code
3- some functions should still become static (see comments)
4- at the begining of nxp_p1021.c there is a vague comment "TODO - Ethernet" - please expand the todo with a sentence explaining what's missing, or remove the comment

**Comment by dgarske** [comment]
> 1- OK keeping the CRC code but please change if 0 into something more meaningful
> 2- Eth initialization: there is a "commented out" memcpy under `#if 0`, please use a meaningful define or eliminate dead code
> 3- some functions should still become static (see comments)
> 4- at the begining of nxp_p1021.c there is a vague comment "TODO - Ethernet" - please expand the todo with a sentence explaining what's missing, or remove the comment

These issues are all fixed. Your comment says it was on "outdated". Please make sure you are reviewing the latest version.

**Review by danielinux — APPROVED** [review]

**Pull request merged**